### ISSUE TITLE
Define a plan for HTTP engine API

### ISSUE THREAD

**Issue by vieux** [moby/moby#32873]
The docker CLI is moving to it's own repository, `docker/cli` See https://github.com/moby/moby/pull/32694

The current plan is to leave the HTTP engine API in the moby repo. It will work for now because "docker engine" is still a monolith, but it will get problematic as soon as we start splitting it into components.

We need a plan moving forward. Suggestions welcome :D

**Comment by cpuguy83** [comment]
I suggest the HTTP API be a component itself that just wraps the GRPC APIs.
Easier said than done, but basically this would need to be adapted over time such that API's either map directly to the monolith or to the external component(s) as they are added.

**Comment by duglin** [comment]
+1 to @cpuguy83's comment. We don't want to impact those users who want to talk to Docker/Moby via curl.

**Comment by cpuguy83** [comment]
@duglin To be clear, I do not think we should not add new features to the HTTP API, only adapt it to ensure existing functionality works against the split components.

**Comment by duglin** [comment]
@cpuguy83 the double negative is throwing me :-)  As we move forward I think supporting HTTP at some level (whether its docker or moby) will be important. Not everyone will want to produce a grpc client to interact with the backend.

**Comment by cpuguy83** [comment]
> the double negative is throwing me
grr... started the sentance one way and finished it another... "we should not add new features".

The goal is to provide SDK's for all (as many as possible) languages.

**Comment by duglin** [comment]
Just my 2 cents, but I think that's chopping off a non-zero size community of folks who want to do bash or simple http interactions w/o needing to pull down additional dependencies/libraries.

**Comment by rogaha** [comment]
@cpuguy83 maybe something like https://github.com/grpc-ecosystem/grpc-gateway would help to keep backwards compatibility, allowing users to still consume the APIs using `HTTP`.  

![68747470733a2f2f646f63732e676f6f676c652e636f6d2f64726177696e67732f642f3132687034435071724e5046686174744c5f63496f4a707446766c41716d35774c513067677149356d6b43672f7075623f773d37343926683d333730](https://cloud.githubusercontent.com/assets/4483394/25487189/4df10e36-2b18-11e7-9e46-48e19ce14125.png)

**Comment by cpuguy83** [comment]
@duglin I think we need to look at it as supporting the v1 API while also adding a v2 API.

**Comment by duglin** [comment]
Someone once told me there's a way to tell grpc to support vanilla http in addition to the more optimized grpc path - haven't had a chance to look into it, but if so perhaps its as simple as just asking grpc for it.

**Comment by duglin** [comment]
@cpuguy83 yep. While it would be nice if going forward the existing API grew instead of being totally revamped (ie. replaced with a v2), I think treating today's API as (sort of) frozen while moving forward with a v2 if needed due to the other changes going on is workable. My bigger concern was dropping vanilla HTTP support regardless of its exact shape.

**Comment by dnephin** [comment]
There is some discussion on the forums in this topic: https://forums.mobyproject.org/t/proposal-engine-v1-rest-api-transition/

I forgot to post here when I posted the proposal.

**Comment by mavenugo** [comment]
This discussion (https://forums.mobyproject.org/t/topic-find-a-good-and-non-confusing-home-for-the-remaining-monolith/37/68) adds more color to this topic.
